### PR TITLE
API extension to support  user-selectable network  ports

### PR DESCRIPTION
**Pull request by fduncanh**
Since Waester's  pull request #196   hasn't been worked on to implement the command line option that FD- and pallas requested,
I have done it in this pull request.

(the original pull request is reworked, as the ports can now be changed with the command line option -p[n].)

-p uses the hard coded legacy ports found  by Waester.
-pn  uses both tcp and udp ports n, n+1, n+2 in range 1024- 65535  

**This is now implemented through an addition to  the API of the airplay  RAOP library  to allow port variables  to be
selected**

The help option and documentation are also updated.      tested by mirroring an iPad running iPadOS 14.6.

### CODE REVIEW

**Comment by fduncanh** [comment]
updated to expand the  -sp option so -sp activates Waester's hard-coded legacy port choices, and -sp[n] (e.g. -sp34567) changes the static ports to tcp AND udp in range n-(n+2)  (e.g. 34567-34569  , 3 udp and 3 tcp ports). where ports are in the "ephemeral port" range 32768-65535,

**Comment by FD-** [comment]
Thanks for your contribution! Here are a few remarks I have regarding your changes:

- I think a cleaner solution for setting ports would be passing them (or the first in a sequential series) as arguments to the init functions (0 to pick randomly, but that comes naturally anyway). That way, the different components would not have to deal with the port decision policy at all. What do you think?
- Please fix your indentation (you used tabs in a few places) and put a space char between if and the condition.

**Comment by fduncanh** [comment]
Hi FD,  I  redid the structure in a clearer way.

-p  shows ports to be used and stops
 default ports are now Waester's set of  six (3 tcp, 3 udp) "legacy" hard coded ports
-pn   changes ports to tcp n n+1 n+2 udp n n+1 n+2 in range 32768-65535
-pn -p  shows the ports that  -pn selects.

I go back and check for tabs etc

**Comment by fduncanh** [comment]
I thing the tabs are all gone, 
"if"'s followed by space etc

**Comment by fduncanh** [comment]
code is cosmetically "squeeky clean" now, and tested. ready to merge.

**Comment by FD-** [comment]
Thanks for the rework! Still, I'd prefer a solution where ports are passed as arguments to the init functions (0 to pick randomly, but that comes naturally anyway), so that we don't need a single static variable, which defeats the purpose of the class-like struct.

**Comment by fduncanh** [comment]
FD- could you be more explicit?   what class-like struct are you referring to? what is the "single static variable"

 right now,  the six (in fact five) non-static port numbers are created in rpiplay.cpp, and then propagated  to the raop_ntp.c (one value), raop_rtp.c   (two values) raop_rtp_mirror.c (one value) by declaring them there as "extern".      

Are you asking me  to create a new struct to contain pointers to the port values and make then available that way ?,
is there already a suitable one to use?

Or is it a problem that    a single variable n is used to set up  ports tcp n-(n+2), udp n-(n+2).   it would be easy to set up a more cumbersome but flexible scheme in rpiplay.cpp

or is it  the use of extern to pass the port  values that you dont like? (probably a bit archaic)

**Comment by fduncanh** [comment]
since raop.h  includes raop_ntp. h,  the #include "raop_ntp.h" in raop_ntp.c can be replaced by #include "raop.h" there.
(I  checked this does not seem to break anything)

so certainly a struct containing the port declarations in rpiplay.cpp can be placed in raop.h  instead.

struct network_ports_s {
    unsigned short RAOP_DNS_TCP_PORT         = 7000;
    unsigned short AIRPLAY_DNS_TCP_PORT      = 7001; /* will always be RAOP_DNS_TCP_PORT + 1*/
    unsigned short MIRROR_DATA_TCP_PORT      = 7100; /* Legacy value needed for IOS6 (or older) clients*/
    unsigned short AUDIO_DATA_UDP_PORT       = 6000;
    unsigned short AUDIO_CONTROL_UDP_PORT    = 6001;
    unsigned short VIDEO_TIMING_UDP_PORT     = 7011;
}

and the external declarations in raop_ntp.c etc could removed.    This would be a cosmetic change.

are you asking for something more substantial?

**Comment by fduncanh** [comment]
FD-, I'm not sure I understand what you want, but I will look into it.   I'tt create a struct with pointers to the port values.

**Comment by FD-** [comment]
Sorry, I should have explained in detail. What I'm suggesting is placing
the port variables inside the individual structs (raop_ntp, raop_*) instead
of having them as separate globals. The library in theory supports creating
multiple instances of these components, which is no longer possible if the
port is a single global variable (per struct type).

In object-oriented terminology, I want you to make the ports instance
fields instead of static class fields.

The field inside each struct would then be filled from a parameter to the
respective init function.

Hope this makes sense! Let me know if you have further questions!


fduncanh ***@***.***> schrieb am Mi., 16. Juni 2021, 08:50:

**Comment by fduncanh** [comment]
**The field inside each struct would then be filled from a parameter to the
respective init function.**

But this seems to be what the code already does now?  I'm not understanding your point.

the struct e.g. raop_ntp_indeed contains "unsigned short  timing_port"
which  is given a value  when raop_ntp_init_socket is called.  

right now the non-static variable "tport" in raop_ntp_init_socket is reset to 0 each time the function is called.

&tport is passed to netutils_init_socket.   If the callling value of tport is invalid, it is overwritten with a valid value,
and if the initial valid is valid it is retained.   The return value is  the address of the local timing port  socket. 

on return from netutils_init_socket, the value of tport is copied into the struct raop_ntp_s  member "timing_lport".

What I am doing is replacing  the initial value non-static value 0 of tport with the value of a global variable VIDEO_TIMING_UDP_PORT

I dont think raop_ntp_init_socket is ever called more than once.    If that is a worry, I could always set the global variable
VIDEO_TIMING_UDP_PORT to zero after the first call, so any subsequent attempt to open a port would get a random result.

File: lib/raop_ntp.c [hunk 1]
```diff
@@ -21,7 +21,7 @@
 #include <string.h>
 #include <stdbool.h>
 
-#include "raop_ntp.h"
```
Comment (FD-) [inline lib/raop_ntp.c:24]: Why did you remote that include?

**Comment by FD-** [comment]
I now had a closer look through the existing codebase (it's been a while since I worked on the project). The details I describe below might differ from what I wrote above, but the idea remains the same.

My main point is: The library currently supports multiple connections at once (even if the renderers don't support that), each with its own instance of raop_ntp, raop_rtp and raop_rtp_mirror. These are in the raop_conn_s structs. I'd like to retain that architecture, which means that each individual connection needs its own set of ports for raop_ntp, raop_rtp and raop_rtp_mirror.

AFAICT, your current design doesn't cover that.

Additionally, I don't like how you currently have your declarations and implementations all over the place, breaking the conceptual separation between the library and the program. Generally, please keep the declaration in the header corresponding to the file that implements the function, otherwise we just get a huge mess over time.

File: lib/raop_ntp.c (hunk 1 above)
Comment (fduncanh) [reply lib/raop_ntp.c:24]: i reversed this, and added the new declartations in a new include file raop_ports.h, include here and in raop.h

File: lib/raop.h [hunk 2]
```diff
@@ -26,6 +26,13 @@ extern "C" {
 #define RAOP_LOG_INFO        6       /* informational */
 #define RAOP_LOG_DEBUG       7       /* debug-level messages */
 
+#define UDP_DATA_PORT        1
+#define UDP_CONTROL_PORT     2
+#define UDP_TIMING_PORT      3
+#define TCP_MIRROR_PORT      4
+#define TCP_RAOP_PORT        5
+
+void initialize_port (unsigned short *, unsigned short);
```
[the remark this answers is not in the document]
Comment (fduncanh) [reply lib/raop.h]: moved to a new include "raop_port.h"

**Comment by fduncanh** [comment]
I believe that the change I  proposed maintains the separation between the library and the rpiplay.cpp program.

From the library's viewpoint, in my implementation it just knows about a "user-supplied" function with a prototype

```void  initialize_port(unsigned short *port, unsigned short port_id)```

(now)  defined in  a new include **raop_port.h**  , where \"port_id\"  takes one of 5 values also defined in raop_port.h 
This is included in raop_ntp.c and in raop.     Here the function is supplied in rpiplay.cpp. 

The only change to the current header files is now the inclusion of raop_port.h in raop.h.  (I reversed the previous change in 
raop_ntp.c, which now again includes raop_ntp.h and not raop.h)

**Comment by fduncanh** [comment]
Looking at this some more,   maybe it is best to add the proposed set of five ports calculated from user input to
the main proram (rpiplay) as  an input argument to start__sever e.g. &network_ports,  a struct of 5 port numbers.

**Comment by fduncanh** [comment]
finally, I think I give you a clean implementation:  the five initial port values are calculated from user input in main, where they are placed in an array which is passed by address to **start_server()** as a new input variable, and immediately copied to an internal array variable **port_t raop_ports** (defined in raop_ports.h) that is accessed by the init functions.    

I  think this meets your your requirement __"I'd like to retain that architecture, which means that each individual connection needs its own set of ports for raop_ntp, raop_rtp and raop_rtp_mirror"__.

**Comment by fduncanh** [comment]
quoting FD-:

_My main point is: The library currently supports multiple connections at once (even if the renderers don't support that), each with its own instance of raop_ntp, raop_rtp and raop_rtp_mirror. These are in the raop_conn_s structs. I'd like to retain that architecture, which means that each individual connection needs its own set of ports for raop_ntp, raop_rtp and raop_rtp_mirror._







FD-  I finished doing exactly this,  this weekend.   As you suggested, I managed to understand how to properly inject the ports  requested into the  int structures.   I had to work out how to store the port requests  in the structure raop_t so they could be injected into raop_ntp_t raop_rtp_t and raop_mirror_rpt_t when the handle calls the init functions..   

The API now contains  functions for writing the requested port-values into raop_t immediately after it has been created by raop_init  (but before start_raop is called ) in start_server.    The API now contains set_timing_lport etc in addition to set_port,


The changes are tested and work.

**Comment by fduncanh** [comment]
This got automatically closed when I reset  my repo for a fresh start.
The code written as  FD- requested is in a new pull request.

**Pull request closed without merging**